cumulusci/tasks/bulkdata.py

```diff
@@ -797,6 +799,313 @@ def _sqlite_dump(self):
+                                    obj in self.mapping_objects
+                                    and target in self.mapping_objects
+                                ):
+                                    if obj not in self.refs:
+                                        self.refs[obj] = set()
+                                    self.refs[obj].add(target)
+
+    def _build_mapping(self):
+        objs = set(self.schema.keys())
+        stack = self._split_dependencies(objs, self.refs)
+
+        field_sort = (
+            lambda f: "  " + f
+            if f == "Id"
+            else (" " + f if f in self.core_fields else f)
+        )
+
+        self.mapping = OrderedDict()
+        for obj in stack:
+            key = "Insert {}".format(obj)
+            self.mapping[key] = OrderedDict()
+            self.mapping[key]["sf_object"] = "{}".format(obj)
+            self.mapping[key]["table"] = "{}".format(obj.lower())
+            fields = []
+            lookups = []
+            for field in self.schema[obj].values():
+                if field["type"] == "reference":
+                    lookups.append(field["name"])
+                else:
+                    fields.append(field["name"])
+            self.mapping[key]["fields"] = OrderedDict()
+            if fields:
+                if "Id" not in fields:
+                    fields.append("Id")
+                fields.sort(key=field_sort)
+                for field in fields:
+                    self.mapping[key]["fields"][field] = (
+                        field.lower() if field != "Id" else "sf_id"
+                    )
+            if lookups:
+                lookups.sort(key=field_sort)
+                self.mapping[key]["lookups"] = OrderedDict()
+                for field in lookups:
+                    self.mapping[key]["lookups"][field] = {
+                        "table": self.schema[obj][field]["referenceTo"][0].lower()
+                    }
+                    if len(self.schema[obj][field]["referenceTo"]) > 1:
+                        self.logger.warning(
+                            "Field {}.{} is a polymorphic lookup, which is not supported".format(
+                                obj, field
+                            )
+                        )
+
+    def _run_task(self):
+        self._collect_objects()
+        self._build_schema()
+        self._build_mapping()
+        with open(self.options["path"], "w") as f:
+            yaml.safe_dump(self.mapping, f)
+
+    def _split_dependencies(self, objs, dependencies):
+        stack = []
+        objs_remaining = objs.copy()
+
+        # The structure of `dependencies` is:
+        # key = object, value = set of objects it references.
+
+        # Iterate through our list of objects
+        # For each object, if it is not dependent on any other objects, place it at the end of the stack.
+        # Once an object is placed in the stack, remove dependencies to it (they're satisfied)
```

Comment:
So this is basically a topological sort of the object dependency graph?

Reply:
Basically, yeah, accounting for the fact that it's not a DAG.